pypeit/spectrographs/keck_kcwi.py

```diff
@@ -607,51 +617,251 @@ def bpm(self, filename, det, shape=None, msbias=None):
         for bb in range(len(bc)):
             bpm_img[bc[bb][2]:bc[bb][3]+1, bc[bb][0]:bc[bb][1]+1] = 1
 
-        return bpm_img
+        return np.flipud(bpm_img)
+
+    @staticmethod
+    def is_nasmask(hdr):
+        """Determine if a frame used NAS
 
-    def set_wcs(self, hdr):
-        """Set the WCS for this spectrograph
+        Parameters
+        ----------
+        hdr : fits header
+            The header of the raw frame.
+
+        Returns
+        -------
+        bool : True if NAS used
         """
-        # TODO :: Need to set all of these!!
-        msgs.error("Should be using astropy.wcs")
-        ra = None
-        dec = None
-        wave0 = None
-        crpix1 = None
-        crpix2 = None
-        crpix3 = None
-        cd11 = None
-        cd21 = None
-        cd12 = None
-        cd22 = None
-        dwout = None
-        # WCS keywords
-        hdr['WCSDIM'] = (3, 'number of dimensions in WCS')
-        hdr['WCSNAME'] = ('KCWI', 'Name of WCS')
-        hdr['EQUINOX'] = (2000, 'EQUINOX')
-        hdr['RADESYS'] = ('FK5', 'WCS system')
-        hdr['CTYPE1'] = ('RA---TAN', '')
-        hdr['CTYPE2'] = ('DEC--TAN', '')
-        hdr['CTYPE3'] = ('AWAV', 'Air Wavelengths')
-        hdr['CUNIT1'] = ('deg', 'RA units')
-        hdr['CUNIT2'] = ('deg', 'DEC units')
-        hdr['CUNIT3'] = ('Angstrom', 'Wavelength units')
-        hdr['CNAME1'] = ('KCWI RA', 'RA name')
-        hdr['CNAME2'] = ('KCWI DEC', 'DEC name')
-        hdr['CNAME3'] = ('KCWI Wavelength', 'Wavelength name')
-        hdr['CRVAL1'] = (ra, 'RA zeropoint')
-        hdr['CRVAL2'] = (dec, 'DEC zeropoint')
-        hdr['CRVAL3'] = (wave0, 'Wavelength zeropoint')
-        hdr['CRPIX1'] = (crpix1, 'RA reference pixel')
-        hdr['CRPIX2'] = (crpix2, 'DEC reference pixel')
-        hdr['CRPIX3'] = (crpix3, 'Wavelength reference pixel')
-        hdr['CD1_1'] = (cd11, 'RA degrees per column pixel')
-        hdr['CD2_1'] = (cd21, 'DEC degrees per column pixel')
-        hdr['CD1_2'] = (cd12, 'RA degrees per row pixel')
-        hdr['CD2_2'] = (cd22, 'DEC degrees per row pixel')
-        hdr['CD3_3'] = (dwout, 'Wavelength Angstroms per pixel')
-        hdr['LONPOLE'] = (180.0, 'Native longitude of Celestial pole')
-        hdr['LATPOLE'] = (0.0, 'Celestial latitude of native pole')
-        hdr['HISTORY'] = "  {0:s} {1:s}".format(kgeom.progid, systime(0, kgeom.timestamp))
-        hdr['HISTORY'] = "  {0:s} {1:s}".format(pre, systime(0))
-        return hdr
+        if 'Mask' in hdr['BNASNAM']:
+            return True
+        else:
+            return False
+
+    @staticmethod
```

Comment:
Are you adding this because it changes from frame to frame?
Put another way, why not use the `meta` approach?
Note that X-Shooter has frames with multiple binnings.

Reply:
This was lazy of me... should have realised that was a better approach - fixed throughout!